### PR TITLE
The contextual rules check the agent ID

### PR DESCRIPTION
**Pull request by carlocas**
|Related issue|
|---------|
[3865](https://github.com/wazuh/wazuh/issues/3865)

## Description

The contextual rules don't check the ID of the agent or manager that generates the event. This results in that some rules are fired when multiples agents send an event when these rules should only be fired when the same agent sends the event. To remediate this issue, I introduced a condition when we check the timeframe of the alerts. The condition is that the agent ID has to be the same that fired the event in the first place. 
In addition, a new label has been added to preserve the previous behavior.  This label is `<global_frequency/>`.

## Tests

- Compilation without warnings in every supported platform
  - [x] Linux
  - [x] Windows
  - [x] MAC OS X
- [x] Source installation
- [ ] Package installation
- [x] Source upgrade
- [ ] Package upgrade
- [x] Custom test
   - [x] Fired an alert of an SSH attack. Check if the alerts 5712 is fired when we introduced the wrong user eight times in a row. 

```
** Alert 1567770100.745521: - syslog,sshd,authentication_failures,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gdpr_IV_35.7.d,gdpr_IV_32.2,hipaa_164.312.b,nist_800_53_SC.7,nist_800_53_AU.3.1,nist_800_53_IA.10,
2019 Sep 06 13:41:40 (centos7) 172.16.1.13->/var/log/secure
Rule: 5712 (level 10) -> 'sshd: brute force trying to get access to the system.'
Src IP: 172.16.1.14
Src Port: 58150
Sep  6 11:41:40 localhost sshd[7260]: Invalid user centos7 from 172.16.1.14 port 58150
Sep  6 11:41:39 localhost sshd[7258]: Invalid user centos7 from 172.16.1.14 port 58148
Sep  6 11:41:38 localhost sshd[7256]: Invalid user centos7 from 172.16.1.14 port 58146
Sep  6 11:41:38 localhost sshd[7254]: Invalid user centos7 from 172.16.1.14 port 58144
Sep  6 11:41:37 localhost sshd[7252]: Invalid user centos7 from 172.16.1.14 port 58142
Sep  6 11:41:36 localhost sshd[7250]: Invalid user centos7 from 172.16.1.14 port 58140
Sep  6 11:41:36 localhost sshd[7248]: Invalid user centos7 from 172.16.1.14 port 58138
Sep  6 11:41:32 localhost sshd[7246]: Invalid user centos7 from 172.16.1.14 port 58136
```
   - [x] Fired an alert of an SSH attack from two different agents. Check if the alerts 5712 is fired only when an agent reaches eight users wrong.
```
** Alert 1568110751.279076: - syslog,sshd,authentication_failures,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gdpr_IV_35.7.d,gdpr_IV_32.2,
2019 Sep 10 12:19:11 (centos7) 172.16.1.13->/var/log/secure
Rule: 5712 (level 10) -> 'sshd: brute force trying to get access to the system.'
Src IP: 172.16.1.14
Src Port: 42264
Sep 10 10:19:10 centos7 sshd[6584]: Invalid user centos7-cliente from 172.16.1.14 port 42264
Sep 10 10:19:09 centos7 sshd[6582]: Invalid user centos7-cliente from 172.16.1.14 port 42262
Sep 10 10:19:04 centos7 sshd[6579]: Invalid user centos7-cliente from 172.16.1.14 port 42260
Sep 10 10:19:04 centos7 sshd[6577]: Invalid user centos7-cliente from 172.16.1.14 port 42258
Sep 10 10:18:59 centos7 sshd[6573]: Invalid user centos7-cliente from 172.16.1.14 port 42256
Sep 10 10:18:58 centos7 sshd[6571]: Invalid user centos7-cliente from 172.16.1.14 port 42254
Sep 10 10:18:57 centos7 sshd[6569]: Invalid user centos7-cliente from 172.16.1.14 port 42252
Sep 10 10:18:56 centos7 sshd[6567]: Invalid user centos7-cliente from 172.16.1.14 port 42250

** Alert 1568110755.280136: - syslog,sshd,authentication_failures,pci_dss_11.4,pci_dss_10.2.4,pci_dss_10.2.5,gdpr_IV_35.7.d,gdpr_IV_32.2,
2019 Sep 10 12:19:15 (centos7-cliente) 172.16.1.14->/var/log/secure
Rule: 5712 (level 10) -> 'sshd: brute force trying to get access to the system.'
Src IP: 172.16.1.13
Src Port: 51236
Sep 10 10:19:14 centos7-cliente sshd[6314]: Invalid user centos7 from 172.16.1.13 port 51236
Sep 10 10:19:07 centos7-cliente sshd[6310]: Invalid user centos7 from 172.16.1.13 port 51234
Sep 10 10:19:02 centos7-cliente sshd[6306]: Invalid user centos7 from 172.16.1.13 port 51232
Sep 10 10:19:01 centos7-cliente sshd[6304]: Invalid user centos7 from 172.16.1.13 port 51230
Sep 10 10:18:53 centos7-cliente sshd[6298]: Invalid user centos7 from 172.16.1.13 port 51228
Sep 10 10:18:52 centos7-cliente sshd[6296]: Invalid user centos7 from 172.16.1.13 port 51226
Sep 10 10:18:51 centos7-cliente sshd[6294]: Invalid user centos7 from 172.16.1.13 port 51224
Sep 10 10:18:49 centos7-cliente sshd[6292]: Invalid user centos7 from 172.16.1.13 port 51222
```

- Memory tests for Linux
  - [x] Scan-build report
  - [x] Valgrind (memcheck and descriptor leaks check)
- Memory tests for Windows
  - [x] Scan-build report

### CODE REVIEW

File: src/headers/rules_op.h [hunk 1]
```diff
@@ -32,6 +32,7 @@
 #define NOT_SAME_SRCIP      0xffd /* 0xfff - 0x002 */
 #define NOT_SAME_ID         0xffb /* 0xfff - 0x004 */
 #define NOT_SAME_AGENT      0xff7 /* 0xfff - 0x008 */
+#define NOT_SAME_AGENT_ID   0xff3
```
Comment (crolopez) [inline src/headers/rules_op.h]: We use the inverse masks when we want to remove some flag activated, or simply to confirm that some check is not going to be evaluated. For example, `NOT_SAME_ID` deactivates `SAME_ID` at bit-level. 
```
  0000 0000 0100 <---- `SAME_ID`
& 1111 1111 1011 <---- `NOT_SAME_ID`
```

Taking this into account, and following the succession, `NOT_SAME_AGENT_ID` (`GLOBAL_FREQUENCY`) should take the value `0x800` and should be below `NOT_SAME_FIELD`.

**Comment by carlocas** [comment]
### Issue Update

The fields `NOT_SAME` (user, srcip, id, and agent) have been deprecated. These fields only add complexity to the code.

File: src/analysisd/eventinfo.c [hunk 2]
```diff
@@ -76,6 +76,10 @@ Eventinfo *Search_LastSids(Eventinfo *my_lf, RuleInfo *rule, __attribute__((unus
             goto end;
         }
 
+        if (!(rule->context_opts & GLOBAL_FREQUENCY) && (strcmp(lf->agent_id, my_lf->agent_id))){
```
Comment (chemamartinez) [inline src/analysisd/eventinfo.c]: Check the existence of `lf->agent_id` and `my_lf->agent_id` to avoid a crash. The same for the other occurrences.

**Comment by chemamartinez** [comment]
Tested last changes by crossing events from several agents at the same time, now, the `frequency` alert is fired for each agent independently, nice work @carlocas.

**Comment by branchnetconsulting** [comment]
When documenting the new `<global_frequency/>`  label, it would be good to make it clear that the frequency is only global at the manager level, not at the manager cluster level.  We can't keep track of composite rule event counters at the level of the manager cluster yet.  The word "global" could give that impression if we are not quite clear about that.

**Comment by chemamartinez** [comment]
> When documenting the new `<global_frequency/>` label, it would be good to make it clear that the frequency is only global at the manager level, not at the manager cluster level. We can't keep track of composite rule event counters at the level of the manager cluster yet. The word "global" could give that impression if we are not quite clear about that.

Hi @branchnetconsulting,

You are right, I added a note to clarify that point: https://github.com/wazuh/wazuh-documentation/commit/53ed88be3ff4ac36da1536af518a206626742720

Regards.

**Pull request merged**